### PR TITLE
Introduce a way to mix generated datasets before sending to training

### PR DESCRIPTION
**Pull request by bbrowning**
See #162 

Rebased from aakankshaduggal#4,  aakankshaduggal#13, and aakankshaduggal#14 reducing scope and splitting what can out into subsequent follow-up PRs for the overall data mixing functionality instead of landing everything as one big PR.

Refactored and changed by me to be as least disruptive as possible in merging in, which means leaving the legacy train and messages outputs in place in generated_data.py.

This introduces a new Recipe class that references multiple generated datasets and has logic to combine those into a single mixed dataset based on a configurable number / ratio of samples to take from each dataset.

This adds new output artifacts for each data generation run:

* a `node_datasets_*` subfolder that contains the raw generated samples   from each taxonomy leaf node as a set of jsonl files - within here,  each skill taxonomy node gets a single jsonl and knowledge nodes get   two jsonl files used in separate downstream steps of training
* a `skills_train_msgs_*.jsonl` and a `knowledge_train_msgs_*.jsonl`  file that contains mixtures of the above raw generated samples based  on proportions specified during the mixing process

```
Co-authored-by: shivchander <shivchander.s30@gmail.com>
Co-authored-by: Khaled Sulayman <khaled@thesulaymans.com>
Co-authored-by: abhi1092 <abhi1092@gmail.com>
Co-authored-by: Aakanksha Duggal <aduggal@redhat.com>
```

### CODE REVIEW

**Comment by bbrowning** [comment]
The e2e test finishes synthetic data generation now, although fails during training because of `/tmp/tmp.JJXU361JVM/.local/share/instructlab/datasets does not contain training or test files, did you run `ilab data generate`?` This is because the mixed files do not match the `train_*` or `test_*` prefixes that it expects those files to have, and instead have names like `skills_train_msgs_2024-07-18T16_10_49.jsonl` or `knowledge_train_msgs_2024-07-18T16_10_49.jsonl`.

File: src/instructlab/sdg/generate_data.py
Comment (markmc) [inline src/instructlab/sdg/generate_data.py:313]: Is the roundtrip to disk necessary? i.e. we're writing to the dataset to a file just for it to be loaded again shortly after?

The benefit is that we have a record of this intermediate stage? Or reduced memory usage?

File: src/instructlab/sdg/utils/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/utils/datamixing.py:48]: Another place where we should be using `PipelineContext.num_proc`

File: src/instructlab/sdg/generate_data.py
Comment (markmc) [inline src/instructlab/sdg/generate_data.py:199]: I feel like the main purpose of this file is to specify some initial data sets that should be mixed with the generated datasets?

Couldn't that be specified in the pipeline config:

```
version: "1.0"
initial_datasets:
  - path: ../../datasets/mydataset.jsonl
    sampling_size: 30
blocks:
  - name: gen_questions
    type: LLMBlock
```

?

It also seems to serve a secondary purpose of writing a record of the datasets that were mixed to produce the final output?

File: src/instructlab/sdg/utils/parse_and_convert.py
Comment (markmc) [inline src/instructlab/sdg/utils/parse_and_convert.py:129]: "kowlege" typo ... and the file doesn't exist

File: src/instructlab/sdg/utils/parse_and_convert.py
Comment (markmc) [inline src/instructlab/sdg/utils/parse_and_convert.py:163]: Unclear to me where `dataset_type` is coming from and what it's valid values are?

**Review by markmc — COMMENTED** [review]
A bunch of comments all over the place, not terribly useful I'm afraid. Will try to follow up with some higher-level comments

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: @aakankshaduggal Can you answer Mark's question about default recipes? They only specify a system prompt, and the skill and knowledge recipes (2 different files) are required to have the identical system prompt. What's the overall purpose these default recipes are serving?

File: src/instructlab/sdg/utils/parse_and_convert.py
Comment (bbrowning) [reply src/instructlab/sdg/utils/parse_and_convert.py:286]: @aakankshaduggal @shivchander Do either of you have a document or description of what exactly Phase 0.7 and Phase 1.0 mean? This sounds both like something we should write into a dev-doc (see Mark's request for that at https://github.com/instructlab/sdg/pull/163#issuecomment-2239071197) as well as something that either needs to make its way into code comments or I can swap those terms out for something more human-friendly in the method names and references here, if there's some common name given to these training phases outside of phase 0.7 and phase 1.0.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:43]: @aakankshaduggal Is there some background on why 30 was chosen as the sampling size for skills recipes? Why not 5, 100, 500? Is this actually a constant that no one should ever modify, or is this something that would need to be tweaked depending on the dataset in use?

File: src/instructlab/sdg/utils/parse_and_convert.py
Comment (bbrowning) [reply src/instructlab/sdg/utils/parse_and_convert.py:129]: This feels like dead/outdated code to me, and I'm tempted to rip out all the auxiliary dataset bits in this PR to let someone else add them later with a follow-up and explanation of what it is and how it's used. I can't see anything that would cause this to actually run today, and it's code not exercised by any of the manual or automated testing I've performed. Looping in @aakankshaduggal and @shivchander in case they feel otherwise and can explain what this does.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:177]: I split the concerns of figuring out how to inject initial data sets into #171, which means for the scope of this PR the default recipes are really just bringing the system prompt. These files don't seem like a great way to manage that, since as you pointed out they have to be equal in both knowledge and skills recipes. I'm giving a chance for some of the others I tagged to speak up here, but if they don't then will proceed with removing the default recipe files entirely from this PR for now.

File: src/instructlab/sdg/generate_data.py
Comment (derekhiggins) [inline src/instructlab/sdg/generate_data.py:502]: Decoding this is difficult, maybe something like
```
if not (len(sdgs) and len(sdgs[0].pipelines)):
    assert (.......
```

File: src/instructlab/sdg/generate_data.py
Comment (derekhiggins) [inline src/instructlab/sdg/generate_data.py:546]: I'm missing context here but I'd love to know the signifigance of p07 and p10 below

**Review by derekhiggins — COMMENTED** [review]
I think I need to see the e2e runthough to  understand whats going on here

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:38]: This is what the prompt was changed to in the research fork, although there they are loading it from default_recipe.yaml files that I've moved out of this PR just for simplicity sake for now. However, it was hardcoded in those files to this value.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:578]: That's not a bad idea - I wasn't completely happy with this, as it feels like an awkward way to get a context back from `_sdg_init`. A prior iteration of this just modified `_sdg_init` to also return a PipelineContext, but then that felt odd in a different way because the existing return value is a list of Pipeline objects. It may be cleaner to create a new method that generates a PipelineContext, call that from `generate_data`, and then pass that context into `_sdg_init` instead of having `_sdg_init` create the PipelineContext itself.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: I don't have full context on why they're named this, but did push some additional changes that attempts to document what's in these p07 and p10 datasets at least.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:38]: @aakankshaduggal Do you know if this needs to be conditional and configurable? It was hardcoded into the recipe files before - does that mean it was generally just expected to be hardcoded? Or was there an expectation that an out-of-band process might be replacing those recipes files on disk before running data generation, thus overwriting this system prompt?

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:47]: Ok, I documented this constant to the best of my current understanding.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:301]: I've refactored this PR to be as purely additive as possible, including leaving the previous code in place that generates the `train_*.jsonl` files and `messages_*.jsonl` files. So, we're no longer removing that file from our output. This leaves the door open to a later discussion and follow-up about deprecating/removing those legacy file outputs separately from adding in the new mixed datasets.

File: src/instructlab/sdg/generate_data.py
Comment (markmc) [reply src/instructlab/sdg/generate_data.py:578]: See https://github.com/instructlab/sdg/pull/180

```
-     sdg_knowledge, sdg_freeform_skill, sdg_grounded_skill = _sdg_init(
+     ctx, sdg_knowledge, sdg_freeform_skill, sdg_grounded_skill = _sdg_init(
         pipeline,
         client,
         model_family,
```

Comment (markmc) [reply src/instructlab/sdg/generate_data.py:578]: but passing a PipelineContext to 'sdg_init()' would be fine too

Comment (markmc) [reply src/instructlab/sdg/generate_data.py:578]: Changed #180 to do what you suggested, thanks

https://github.com/instructlab/sdg/commit/57b2a77bdbcdf8b15a6d3453298f3b2854c659df

File: src/instructlab/sdg/utils/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/utils/datamixing.py]: Agreed, will bring this up one level and ensure private methods use the leading underscore naming convention to control what what we expose from here.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: I split writing recipe files to disk out as #185 to reduce scope of this PR.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: My understanding is that it's so we have an artifact trail of the raw generated data as well as the mixed data, and when #185 goes in the recipe files will reference this raw generated data to show what was mixed into the overall mixed dataset. The mixing process can reduce the number of samples from a given taxonomy leaf in the overall mixed dataset, so these intermediate files also allow the entire generated output for each leaf node to be retained as an artifact.

Reduce memory usage is the other side-effect of this, since we're only ever keeping the dataset for a single leaf node in memory until the mixing process, where we sample some number/percent of the overall generated samples for each leaf instead of loading the entirety of all leaf node generated samples at once.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: They were not TODOs, but were confusing and so are removed with separate docs added to explain phase 0.7 and 1.0 in their relevant _create_phase methods.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: Agreed, and I'll let #171 handle dealing with configurable system prompts that match the precomputed datasets as support for precomputed datasets is added. For now this was changed to retain the previous behavior of a hardcoded _SYS_PROMPT until/unless we need configurable system prompt for other reasons.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: The logging improvements were removed and can be made in a follow-up PR to reduce scope of this one.

File: src/instructlab/sdg/utils/parse_and_convert.py
Comment (bbrowning) [reply src/instructlab/sdg/utils/parse_and_convert.py]: Outdated now, as the previous _convert_to_messages is now retained which does have a `metadata` column.

**Comment by bbrowning** [comment]
I think I've addressed most of the initial feedback, although still waiting on a couple of answers from others to ensure I've documented things properly and didn't prune too much out in my attempts to streamline this.  I've also parked a copy of my fork's branch at https://github.com/bbrowning/instructlab-sdg/commits/data-mixing-full-history/ up to this point for reference to myself or anyone else that works on follow-up PRs to remove things I'm pruning out of scope for this one.

Now, on to the squash/rebase phase to get the list of commits down to a reasonable number and ensure authorship is correctly attributed. Once that's done, I think this will be in decent shape for additional review and merging consideration, with the understanding that it's not perfect but has to get in under specific time constraints and is a prerequisite for some smaller follow-up PRs around recipe files, system prompts, and removal of legacy train/messages formats as well as overall integration testing end-to-end.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: Resolving this as that code is no longer present in this PR.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: Thanks, rebased now that 180 merged and resolving this as the context is now gotten without this type of hack.

**Comment by bbrowning** [comment]
Ok, I feel like this is in a point where it's good enough to review. I'll be ready to address any concerns quickly to keep round-trip time of that whole process as short as possible. After merging we'll need some follow-up PRs on top of this work (and tracked in #162) to write out recipe yaml files, add auxiliary datasets, add precomputed datasets, and remove the legacy train/messages json output formats.

File: src/instructlab/sdg/generate_data.py
Comment (markmc) [inline src/instructlab/sdg/generate_data.py:638]: This can use `is_knowledge`

File: src/instructlab/sdg/generate_data.py
Comment (markmc) [inline src/instructlab/sdg/generate_data.py:390]: ChatGPT offers 2 possible explanations for what "raft" could mean here:

> Random Augmented Feature Training (RAFT): In machine learning, "RAFT" might stand for Random Augmented Feature Training, a technique where random augmentation is applied to the features of a dataset to improve the model's robustness and performance. The function's description suggests that additional context from random samples is added, which could be a form of data augmentation.
>
> Reference Augmented Fine-Tuning (RAFT): In NLP, RAFT might refer to adding additional context or references to enhance the training data. By incorporating context from other samples, the function could be aiming to provide a more comprehensive background, potentially improving the quality of responses generated by a language model.

Seems like it's the latter? Expanding the acronym and linking to https://aka.ms/raft-paper would help

File: src/instructlab/sdg/generate_data.py
Comment (markmc) [reply src/instructlab/sdg/generate_data.py:648]: Ok, take a look at https://github.com/markmc/sdg/commit/b01f1ca7d8e7af802bf12ab409aed9c0e1267fa4 in https://github.com/markmc/sdg/tree/data-mixing-mixer

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: Yes, good catch! I missed that when rebasing on top of the latest mmlubench changes.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: That keeps generate_data.py much cleaner, nice! I haven't had a chance to pull and run it locally yet, but it looks like we may be missing the final call to `mixer.generate()`? Or I'm just not finding it?

Comment (markmc) [reply src/instructlab/sdg/generate_data.py]: Fixed, thank you!

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py]: I do think this in the spirit of that RAFT paper, but I'll give this a different name that doesn't require understanding the acronym. Something like `_add_extra_contexts_to_samples`  since it's adding extra contexts to each sample in the dataset.

**Comment by markmc** [comment]
Looking into:

```
 File "/home/runner/work/sdg/sdg/venv/lib/python3.11/site-packages/instructlab/sdg/generate_data.py", line 404, in generate_data
    mixer.generate()
  File "/home/runner/work/sdg/sdg/venv/lib/python3.11/site-packages/instructlab/sdg/datamixing.py", line 468, in generate
    self._gen_mixed_data(
...
  File "/home/runner/work/sdg/sdg/venv/lib/python3.11/site-packages/instructlab/sdg/datamixing.py", line 38, in _load_ds
    dataset = load_dataset("json", data_files=path, split="train")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
  File "/home/runner/work/sdg/sdg/venv/lib/python3.11/site-packages/datasets/data_files.py", line 411, in resolve_pattern
    raise FileNotFoundError(error_msg)
FileNotFoundError: Unable to find '/home/runner/work/sdg/sdg/node_datasets_2024-07-24T14_23_22/compositional_skills_extraction_inference_qualitative_e2e-siblings.jsonl'
```

**Comment by markmc** [comment]
> Looking into:
> 
> ```
> FileNotFoundError: Unable to find '/home/runner/work/sdg/sdg/node_datasets_2024-07-24T14_23_22/compositional_skills_extraction_inference_qualitative_e2e-siblings.jsonl'
> ```

I think it's this:

```
-        recipe.add_dataset(output_file_leaf_node, sampling_size)
+        recipe.add_dataset(output_file, sampling_size)
```

**Comment by bbrowning** [comment]
> > Looking into:
> > ```
> > FileNotFoundError: Unable to find '/home/runner/work/sdg/sdg/node_datasets_2024-07-24T14_23_22/compositional_skills_extraction_inference_qualitative_e2e-siblings.jsonl'
> > ```
> 
> I think it's this:
> 
> ```
> -        recipe.add_dataset(output_file_leaf_node, sampling_size)
> +        recipe.add_dataset(output_file, sampling_size)
> ```

I hit this error locally when testing this a minute ago, and making the change above fixes things on my local machine.

File: src/instructlab/sdg/generate_data.py
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:37]: Leaving this hardcoded for now, as a follow-up PR around recipe yaml files will including loading system prompts from those recipe yaml files.

**Review by markmc — APPROVED** [review]
I'm good with merging this as-is when CI passes, I think the cleanups, documentation, refactoring, scope reduction, etc. has gotten to a place which this is much more maintainable going forward. Thanks, Ben!

**Comment by bbrowning** [comment]
Ok, pushing the big green merge button on this for now as it's a prerequisite for a number of other PRs. Thanks for all the great feedback, and will keep cleaning this up with the work that builds on top of this.

**Pull request merged**
